src/ray/object_manager/object_manager.cc

```diff
@@ -788,8 +788,7 @@ bool ObjectManager::ReceiveObjectChunk(const NodeID &node_id, const ObjectID &ob
   ObjectBufferPool::ChunkInfo chunk_info = chunk_status.first;
   if (chunk_status.second.ok()) {
     // Avoid handling this chunk if it's already being handled by another process.
-    std::memcpy(chunk_info.data, data.data(), chunk_info.buffer_length);
```

Comment:
So chunk_info is holding a shared_ptr to the underlying PlasmaBuffer, so it seems that the memcpy should be safe as long as the chunk_info above is in scope.

I'm wondering if this is actually a new bug introduced by this fix: https://github.com/ray-project/ray/pull/16620/files

Here's my understanding:
- There are two refs to the underlying allocation, one from the PlasmaBuffer itself (decremented on destruction), and another from the client.
- AbortCreate() is (1) calling Release(), then (2) Abort(), then create_buffer_state.erase(object_id).
- Prior to 16620, by step (2) the ref-count drops to zero and plasma server can delete the object safely.
- But after 16620, there may be remaining shared_ptr refs to the PlasmaBuffer for the receive. Hence, we could get a Release() call for an already-deleted object.

If the above is correct, then this fix in this PR may be incomplete: you probably also want to remove the shared_ptr from ChunkInfo.

Reply:
Aside: it's not clear to me why we need this special case Abort() path in the first place... ideally lifetime is managed only via the shared_ptr, and we don't have this separate dual reference counting path.

But maybe this is an issue for the refactor.

Reply:
yup, I agree with your understanding is correct. Here is another race condition in the segfaulted raylet.  The following log validates the theory, as even after we released the object and then aborted the object,  there are still references to the object. (which presumably ref by the PlasmaBuffer from the chunkinfo).
```
[2021-07-20 21:33:31,506 D 4550 4575] store.cc:645: Erasing object 0b9920f3ca70541dffffffffffffffffffffffff010000004a000000 with nonzero ref count0b9920f3ca70541dffffffffffffffffffffffff010000004a000000, num bytes in use is now 1524494562
```



```
[2021-07-20 21:33:31,491 D 4550 4582] object_manager.cc:765: ReceiveObjectChunk on 228b63e21bcee0a7436682edf2de1fe6d7164b688467b094f3dc05d8 from 5c16b78c1611b84df7f908a9e718b0925121383f3a853c1424b5e037 of object 0b9920f3ca70541dffffffffffffffffffffffff010000004a000000 chunk index: 0, chunk data size: 5242880, object size: 45940869
[2021-07-20 21:33:31,506 D 4550 4550] pull_manager.cc:188: Deactivating pull for object 0b9920f3ca70541dffffffffffffffffffffffff010000004a000000
[2021-07-20 21:33:31,506 I 4550 4550] object_buffer_pool.cc:171: Not enough memory to create requested object 0b9920f3ca70541dffffffffffffffffffffffff010000004a000000, aborting
[2021-07-20 21:33:31,506 D 4550 4575] store.cc:633: Erasing object: 0b9920f3ca70541dffffffffffffffffffffffff010000004a000000, address: 0x7f1eb9713cc0, size:45940869
[2021-07-20 21:33:31,506 D 4550 4575] store.cc:645: Erasing object 0b9920f3ca70541dffffffffffffffffffffffff010000004a000000 with nonzero ref count0b9920f3ca70541dffffffffffffffffffffffff010000004a000000, num bytes in use is now 1524494562
[2021-07-20 21:33:31,506 D 4550 4550] pull_manager.cc:394: Removing an object pull request of id: 0b9920f3ca70541dffffffffffffffffffffffff010000004a000000
[2021-07-20 21:33:31,506 D 4550 4550] dependency_manager.cc:27: Object 0b9920f3ca70541dffffffffffffffffffffffff010000004a000000 no longer needed
[2021-07-20 21:33:31,507 D 4550 4583] object_buffer_pool.cc:152: Object 0b9920f3ca70541dffffffffffffffffffffffff010000004a000000 aborted due to OOM before chunk 5 could be sealed
```

Reply:
Brainstorming another way to address this problem, perhaps we could have a Buffer state object that is responsible for calling Abort() on destruction? Then, Abort can be integrated into the normal C++ lifecycle management, and we can also avoid holding the mutex while writing to the buffer.